torch/csrc/jit/script/module.h

```diff
@@ -381,120 +393,291 @@ struct TORCH_API Module {
+        size_(c10::nullopt) {
+    if (!recurse && !return_module && Policy::all_slots) {
       size_ = module_.num_slots();
     }
   }
+
+ private:
   Module module_;
-  // only include Slots of the following type
-  c10::optional<EntityType> type_;
+  bool recurse_;
+  bool return_module_;
   // size of this list, cached on first request
   // when we need to filter the slot list
   mutable c10::optional<size_t> size_;
   friend struct Module;
 };
 
+namespace detail {
+
+// slot_iterator_impl always iterate over all the slots in a module,
+// the Policy template argument determines slots should be returned and their
+// types
+struct TORCH_API ModulePolicy {
+  // the type of the value being returned
+  using value_type = Module;
+
+  // the logic for creating the type being returned, given the raw IValue
+  // of that object.
+  static value_type create(
+      const std::vector<detail::SlotCursor>& cursors,
+      IValue v) {
+    return Module(std::move(v).toObject());
+  }
+  // is slot i in typ something that this iterator should return, otherwise,
+  // we skip it.
+  static bool valid(const ClassTypePtr& typ, size_t i) {
+    return typ->is_module(i);
+  }
+  // are we going to return everything? If so, we can optimize the calculate
+  // of the size of the list.
+  static constexpr bool all_slots = false;
+};
+
+struct TORCH_API ParameterPolicy {
+  using value_type = at::Tensor;
+  static value_type create(
+      const std::vector<detail::SlotCursor>& cursors,
+      IValue v) {
+    return std::move(v).toTensor();
+  }
+  static bool valid(const ClassTypePtr& typ, size_t i) {
+    return typ->is_parameter(i);
+  }
+  static constexpr bool all_slots = false;
+};
+
+struct TORCH_API BufferPolicy {
+  using value_type = at::Tensor;
+  static value_type create(
+      const std::vector<detail::SlotCursor>& cursors,
+      IValue v) {
+    return std::move(v).toTensor();
+  }
+  static bool valid(const ClassTypePtr& typ, size_t i) {
+    return typ->getAttribute(i)->isSubtypeOf(TensorType::get()) &&
+        !typ->is_parameter(i);
+  }
+  static constexpr bool all_slots = false;
+};
+
+struct TORCH_API AttributePolicy {
+  using value_type = IValue;
+  static value_type create(
+      const std::vector<detail::SlotCursor>& frames,
+      IValue v) {
+    return v;
+  }
+  static bool valid(const ClassTypePtr& typ, size_t i) {
+    return true;
+  }
+  static constexpr bool all_slots = true;
+};
+
+// take a Policy object, and make a version of it that returns the slot.
+// along with the fully qualified name of that slot. This is used for the named_
+// variants like named_parameters().
+template <typename Policy>
+struct TORCH_API NamedPolicy {
+  using value_type = Named<typename Policy::value_type>;
+  static value_type create(
+      const std::vector<detail::SlotCursor>& frames,
+      IValue v) {
+    std::string name;
+    if (frames.size() == 1) {
+      name = (frames.back().i_ == -1) ? "" : nameFragment(frames.back());
```

Comment:
Do we really have to special case `size==1`? It looks like it can be handled just fine by the loop below.

Reply:
It's to avoid the overhead of allocating a ostringstream and copying the string twice in the very common non-recursive case.